src/MISC/fix_eco_force.cpp

```diff
@@ -0,0 +1,642 @@
+
+ Copyright (2003) Sandia Corporation.  Under the terms of Contract
+ DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+ certain rights in this software.  This software is distributed under
+ the GNU General Public License.
+
+ See the README file in the top-level LAMMPS directory.
+ ------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing authors: Adrian A. Schratt and Volker Mohles (ICAMS)
+------------------------------------------------------------------------- */
+
+#include "math.h"
+#include "string.h"
+#include "stdlib.h"
+#include "mpi.h"
+#include "fix_eco_force.h"
+#include "atom.h"
+#include "update.h"
+#include "domain.h"
+#include "respa.h"
+#include "force.h"
+#include "pair.h"
+#include "neighbor.h"
+#include "neigh_list.h"
+#include "neigh_request.h"
+#include "citeme.h"
+#include "comm.h"
+#include "output.h"
+#include "math_const.h"
+#include "memory.h"
+#include "error.h"
+
+using namespace LAMMPS_NS;
+using namespace FixConst;
+using namespace MathConst;
+
+static const char cite_fix_eco_force[] =
+  "fix eco/force command:\n\n"
+  "@Article{Schratt20,\n"
+  " author = {A. A. Schratt, V. Mohles},\n"
+  " title = {Efficient calculation of the ECO driving force for atomistic simulations of grain boundary motion},\n"
+  " journal = {Computational Materials Science},\n"
+  " volume = {182},\n"
+  " year = {2020},\n"
+  " pages = {109774},\n"
+  " doi = {j.commatsci.2020.109774},\n"
+  " url = {https://doi.org/10.1016/j.commatsci.2020.109774}\n"
+  "}\n\n";
+
+/* ---------------------------------------------------------------------- */
+
+FixECOForce::FixECOForce(LAMMPS *lmp, int narg, char **arg) :
+    Fix(lmp, narg, arg),
+    dir_filename(NULL), order(NULL), nbr(NULL), list(NULL)
+{
+  if (lmp->citeme) lmp->citeme->add(cite_fix_eco_force);
+
+  // get rank of this processor
+  MPI_Comm_rank(world, &me);
+
+  // check for illegal command
+  if (narg != 7) error->all(FLERR, "Illegal fix eco/force command");
+
+  // set fix flags
+  scalar_flag = 1;          // computes scalar
+  global_freq = 1;          // values can be computed at every timestep
+  extscalar = 1;            // scalar scales with # of atoms
+  peratom_flag = 1;         // quantities are per atom quantities
+  size_peratom_cols = 2;    // # of per atom quantities
+  peratom_freq = 1;         //
+
+  // parse input parameters
+  u_0 = atof(arg[3]);
+  sign = (u_0 >= 0.0 ? 1 : -1);
+  eta = atof(arg[4]);
+  r_cut = atof(arg[5]);
+
+  // read reference orientations from file
+  // work on rank 0 only
+  int n = strlen(arg[6]) + 1;
+  dir_filename = new char[n];
+  strcpy(dir_filename, arg[6]);
+  if (me == 0) {
+    char line[IMGMAX];
+    char *result;
+    int count;
+
+    FILE *infile = fopen(dir_filename, "r");
+    if (infile == NULL) error->one(FLERR, "Fix eco/force file open failed");
+    for (int i = 0; i < 6; ++i) {
```

Comment:
we very recently added new classes to simplify reading potential and parameter files (see PR #2113). There is an update to it in PR #2122 . Since that is likely to be merged before this PR, you may want to look into adopting that code for reading in data from a file.  We are working on replacing ad hoc file readers wherever possible with reasonable effort throughout all of LAMMPS, so you could save us some time and effort and make the reading more robust.